cpp/include/patterns/transform_reduce_by_key_e.cuh

```diff
@@ -0,0 +1,92 @@
+ * input properties.
+ * @tparam AdjMatrixColValueInputIterator Type of the iterator for graph adjacency matrix column
+ * input properties.
+ * @tparam EdgeOp Type of the quaternary (or quinary) edge operator.
+ * @tparam T Type of the initial value of the value in each (key, value) pair.
+ * @tparam KeyIterator Type of the iterator for keys in (key, value) pairs.
+ * @param handle RAFT handle object to encapsulate resources (e.g. CUDA stream, communicator, and
+ * handles to various CUDA libraries) to run graph algorithms.
+ * @param graph_view Non-owning graph object.
+ * @param adj_matrix_row_value_input_first Iterator pointing to the adjacency matrix row input
+ * properties for the first (inclusive) row (assigned to this process in multi-GPU).
+ * `adj_matrix_row_value_input_last` (exclusive) is deduced as @p adj_matrix_row_value_input_first +
+ * @p graph_view.get_number_of_local_adj_matrix_partition_rows().
+ * @param adj_matrix_col_value_input_first Iterator pointing to the adjacency matrix column input
+ * properties for the first (inclusive) column (assigned to this process in multi-GPU).
+ * `adj_matrix_col_value_output_last` (exclusive) is deduced as @p adj_matrix_col_value_output_first
+ * + @p graph_view.get_number_of_local_adj_matrix_partition_cols().
+ * @param e_op Quaternary (or quinary) operator takes edge source, edge destination, (optional edge
+ * weight), *(@p adj_matrix_row_value_input_first + i), and *(@p adj_matrix_col_value_input_first +
+ * j) (where i is in [0, graph_view.get_number_of_local_adj_matrix_partition_rows()) and j is in [0,
+ * get_number_of_local_adj_matrix_partition_cols())) and returns a pair of a key and a transformed
+ * value to be reduced.
+ * @param init Initial value to be added to the value in each transform-reduced (key, value) pair.
+ * @param map_key_first Iterator pointing to the first (inclusive) key to be stored in the returned
+ * cuco::static_map (which is local to this process in mulit-GPU).
+ * @param map_key_last Iterator pointing to the last (exclusive) key to be stored in the returned
+ * cuco::static_map (which is local to this process in multi-GPU).
+ * @return cuco::static_map Hash-based map of (key, value) pairs for the keys pointed by
+ * [map_key_first, map_key_last).
+ */
+template <typename GraphViewType,
+          typename AdjMatrixRowValueInputIterator,
+          typename AdjMatrixColValueInputIterator,
+          typename EdgeOp,
+          typename T,
+          typename KeyIterator>
+cuco::static_map<typename std::iterator_traits<KeyIterator>::value_type, T>
+transform_reduce_by_key_e(raft::handle_t const& handle,
+                          GraphViewType const& graph_view,
+                          AdjMatrixRowValueInputIterator adj_matrix_row_value_input_first,
+                          AdjMatrixColValueInputIterator adj_matrix_col_value_input_first,
+                          EdgeOp e_op,
```

Comment:
Don't we need a transform operator and a reduction operator?  I want to transform each edge tuple into a key/value pair.  Then I need to define a reduction operator that takes pairs of key/value pairs and reduces them.

Reply:
You can transform inside an edge operator.

e_op returns (key, value) pair, and by default, the entire set of values with the same key will be summed. If you need to perform other types of reduction, we need to add a custom reduction operator, but if what you need is "sum" reduction of the values, I think this will be sufficient (I guess this is the case for Louvain). I think we will add reduction operator in the future (e.g. thrust reduce has functions not-taking a reduction operator https://thrust.github.io/doc/group__reductions_ga6ac0fe1561f58692e85112bd1145ddff.html and taking a reduction operator https://thrust.github.io/doc/group__reductions_ga5e9cef4919927834bec50fc4829f6e6b.html#ga5e9cef4919927834bec50fc4829f6e6b I plan to do something similar, but I plan to implement the version taking a reduction operator only when it becomes necessary).

Reply:
That's fine.  I was imagining defining the more complex one and having it default to addition, but deferring until we need the more complex one is also a reasonable choice.